### PR TITLE
SMC ABC

### PR DESCRIPTION
**Pull request by aloctavodia**
This is based on #3369, but as we introduced many changes we decided to open a new PR.

### CODE REVIEW

File: pymc3/distributions/distribution.py [hunk 1]
```diff
@@ -35,6 +35,7 @@ def __new__(cls, name, *args, **kwargs):
 
         if isinstance(name, string_types):
             data = kwargs.pop('observed', None)
+            cls.data = data
```
Comment (junpenglao) [inline pymc3/distributions/distribution.py:42]: Didn't know that the data is never saved with the RV...

File: pymc3/distributions/simulator.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        
+        function : function
+            Simulation function defined by the user.
+        *args and **kwargs : 
+            Arguments and keywords arguments that the function takes.
+        """
+
+        self.function = function
+        observed = self.data
+        super(
```
Comment (junpenglao) [inline pymc3/distributions/simulator.py]: Is this the correct style? seems a bit too many line breaks

Comment (aloctavodia) [reply pymc3/distributions/simulator.py]: that's all black's work!

File: pymc3/step_methods/smc_utils.py [hunk 3]
```diff
@@ -78,3 +85,200 @@ def _posterior_to_trace(posterior, variables, model, var_info):
...
+    low_beta = old_beta = beta
+    up_beta = 2.0
+    rN = int(len(likelihoods) * threshold)
+
+    while up_beta - low_beta > 1e-6:
+        new_beta = (low_beta + up_beta) / 2.0
+        weights_un = np.exp((new_beta - old_beta) * (likelihoods - likelihoods.max()))
+        weights = weights_un / np.sum(weights_un)
+        ESS = int(1 / np.sum(weights ** 2))
+        if ESS == rN:
```
Comment (junpenglao) [inline pymc3/step_methods/smc_utils.py:171]: Hmm, I am not sure I understand this part.

File: pymc3/step_methods/smc_utils.py [hunk 4]
```diff
@@ -78,3 +85,200 @@ def _posterior_to_trace(posterior, variables, model, var_info):
...
+                delta = np.round(delta, 0).astype("int64")
+                q_old = q_old.astype("int64")
+                q_new = (q_old + delta).astype("int64")
+            else:
+                delta[discrete] = np.round(delta[discrete], 0)
+                q_new = floatX(q_old + delta)
+        else:
+            q_new = floatX(q_old + delta)
+
+        if ABC:
```
Comment (junpenglao) [inline pymc3/step_methods/smc_utils.py]: There is no other way to do this?

**Review by junpenglao — COMMENTED** [review]
Overall LGTM, leaves some small comments

File: pymc3/step_methods/smc_utils.py (hunk 3 above)
Comment (aloctavodia) [reply pymc3/step_methods/smc_utils.py:171]: The effective sample size can be estimated from the weights, empirical evidence suggest that good results are obtained when ESS = N/2. Here that observation is used to estimate the beta value to be used on the next stage.

File: pymc3/step_methods/smc_utils.py (hunk 4 above)
Comment (aloctavodia) [reply pymc3/step_methods/smc_utils.py]: this comes from the SMC code, not sure why I need to add this...

Comment (aloctavodia) [reply pymc3/step_methods/smc_utils.py]: to be clear I add it when I refactored the SMC code, I may try to revisit this.

File: pymc3/step_methods/smc_utils.py (hunk 3 above)
Comment (junpenglao) [reply pymc3/step_methods/smc_utils.py:171]: right, why `break` when ESS excatly equal to rN is what I not quite get.

**Comment by junpenglao** [comment]
- please run the example notebook so it has output
- please put the notebook under the SMC session in the doc

I put a "Don't Merge" tag - I think we should cut a new release first.

File: pymc3/step_methods/smc_utils.py (hunk 4 above)
Comment (aloctavodia) [reply pymc3/step_methods/smc_utils.py]: fixed

File: pymc3/step_methods/smc_utils.py (hunk 3 above)
Comment (aloctavodia) [reply pymc3/step_methods/smc_utils.py:171]: The ESS is indirectly measuring the uniformity of the weight. The ESS will be higher for a more uniform distribution of weights than for a peaked distribution. For importance sampling (and SMC) we want more or less uniform distribution of weights, otherwise we loose diversity (one or a few particles explains a lot and the other almost nothing). The ESS can range from 1 to the number of samples (N), with the default value of rN = N/2 being in the middle. If you move super slowly from stage to stage (change beta by a very small amount) then you will guarantee that the weights are very close uniform (and ESS close to N) but you will need to take many steps. If you move too fast (and the prior and posterior are very different) then you are at risk of a single particle having all (or most off the weights).  We use to use the covariance of the weights I change it to ESS to explore other definitions of ESS with the hope of finding an improved way to compute beta.

**Comment by aloctavodia** [comment]
Do we have a SMC section for the notebooks? I realized than SMC (and SMC-ABC) are not listed https://docs.pymc.io/api/inference.html#step-methods we should add them.

**Comment by junpenglao** [comment]
There is one under https://docs.pymc.io/nb_examples/index.html in Simulation-based inference

File: pymc3/distributions/simulator.py (hunk 2 above)
Comment (twiecki) [inline pymc3/distributions/simulator.py]: With python3 you can just call `super()`

File: pymc3/distributions/transforms.py [hunk 5]
```diff
@@ -208,12 +219,19 @@ class Interval(ElemwiseTransform):
...
+        self.b_ = b
 
     def backward(self, x):
         a, b = self.a, self.b
         r = (b - a) * tt.nnet.sigmoid(x) + a
         return r
 
+    def backward_val(self, x):
+        a, b = self.a_, self.b_
+        r = (b - a) * 1/(1+np.exp(-x)) + a  ## fix this
```
Comment (twiecki) [inline pymc3/distributions/transforms.py]: spaces around mathematical operators. does it still need to be fixed?

**Review by twiecki — APPROVED** [review]
Looks great! Needs a line in the release-notes.

File: pymc3/distributions/transforms.py (hunk 5 above)
Comment (aloctavodia) [reply pymc3/distributions/transforms.py]: that comment is about something different, haha, we have to remove it! can we use black to format this? Or we should use autopep8?

File: docs/source/notebooks/table_of_contents_examples.js [hunk 6]
```diff
@@ -53,6 +54,5 @@ Gallery.contents = {
     "lda-advi-aevb": "Variational Inference",
     "normalizing_flows_overview": "Variational Inference",
     "gaussian-mixture-model-advi": "Variational Inference",
-    "GLM-hierarchical-advi-minibatch": "Variational Inference",
-    "ODE_parameter_estimation": "Inference in ODE models"
```
Comment (twiecki) [inline docs/source/notebooks/table_of_contents_examples.js]: dropped ODE.

**Comment by aloctavodia** [comment]
I think this is ready to merge. Do you agree? @junpenglao @twiecki

**Comment by ColCarroll** [comment]
(haven't looked closely, but removing the red "Don't Merge" sign)

File: pymc3/distributions/simulator.py (hunk 2 above)
Comment (junpenglao) [reply pymc3/distributions/simulator.py]: @aloctavodia

File: RELEASE-NOTES.md [hunk 7]
```diff
@@ -4,18 +4,21 @@
 
 ### New features
 - Distinguish between `Data` and `Deterministic` variables when graphing models with graphviz. PR [#3491](https://github.com/pymc-defs/pymc3/pulls/3491).
+- Sequential Monte Carlo - Approximate Bayesian Computation step method is now available. The implementation is in an experimental stage and will be further improved.
 
 ## PyMC3 3.7 (May 29 2019)
 
 ### New features
 
+
```
Comment (twiecki) [inline RELEASE-NOTES.md]: remove new lines.

File: pymc3/distributions/simulator.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        """
+
+        self.function = function
+        observed = self.data
+        super().__init__(shape=np.prod(observed.shape), dtype=observed.dtype, *args, **kwargs)
+
+    def random(self, point=None, size=None):
+        """
+        Draw random values from Simulator
+        Parameters
```
Comment (twiecki) [inline pymc3/distributions/simulator.py:26]: new-line

File: pymc3/distributions/simulator.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        ----------
+        point : dict, optional
+            Dict of variable values on which random values are to be
+            conditioned (uses default point if not specified).
+        size : int, optional
+            Desired size of random sample (returns one sample if not
+            specified).
+        Returns
```
Comment (twiecki) [inline pymc3/distributions/simulator.py:34]: new-line before

File: pymc3/distributions/simulator.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+        raise NotImplementedError('Not implemented yet')
+
+    def logp(self, value):
+        """
+        Parameters
+        ----------
+        value : numeric
+            Value for which log-probability is calculated.
+        Returns
```
Comment (twiecki) [inline pymc3/distributions/simulator.py:47]: same

File: pymc3/step_methods/smc_utils.py [hunk 11]
```diff
@@ -78,3 +85,197 @@ def _posterior_to_trace(posterior, variables, model, var_info):
...
+        model = self.model
+        var_info = self.var_info
+        parameters = {}
+        size = 0
+        for var, values in var_info.items():
+            shape, new_size = values
+            value = posterior[size : size + new_size].reshape(shape)
+            if is_transformed_name(var):
+                var = get_untransformed_name(var)
+                value = model[var].transformation.backward_val(value)
```
Comment (twiecki) [inline pymc3/step_methods/smc_utils.py:249]: can't you do `model[var].transformation.backward(value).eval()`?

File: pymc3/step_methods/smc_utils.py [hunk 12]
```diff
@@ -78,3 +85,197 @@ def _posterior_to_trace(posterior, variables, model, var_info):
...
+            low_beta = new_beta
+    if new_beta >= 1:
+        new_beta = 1
+    sj = np.exp((new_beta - old_beta) * likelihoods)
+    weights_un = np.exp((new_beta - old_beta) * (likelihoods - likelihoods.max()))
+    weights = weights_un / np.sum(weights_un)
+    return new_beta, old_beta, weights, np.mean(sj)
+
+
+def logp_forw(out_vars, vars, shared):
```
Comment (twiecki) [inline pymc3/step_methods/smc_utils.py:185]: Don't we have a similar function elsewhere that does this?

File: pymc3/step_methods/smc_utils.py (hunk 11 above)
Comment (aloctavodia) [reply pymc3/step_methods/smc_utils.py:249]: Yes, but this is faster.

Comment (twiecki) [reply pymc3/step_methods/smc_utils.py:249]: Couldn't you compile and store a `theano.function` of `model[var].transformation.backward` to make it even faster? I'd prefer to avoid the redundancy of the transforms.

File: pymc3/step_methods/smc_utils.py (hunk 12 above)
Comment (aloctavodia) [reply pymc3/step_methods/smc_utils.py:185]: I think this is the only one.

File: pymc3/step_methods/smc_utils.py (hunk 11 above)
Comment (aloctavodia) [reply pymc3/step_methods/smc_utils.py:249]: I think this is not redundant, we already have `forward` and `forward_val`, so it seems reasonable to have `backward` and `backward_val`.

File: pymc3/distributions/transforms.py (hunk 5 above)
Comment (twiecki) [reply pymc3/distributions/transforms.py]: black for new files works well I think.

File: pymc3/step_methods/smc_utils.py (hunk 11 above)
Comment (twiecki) [reply pymc3/step_methods/smc_utils.py:249]: Alright then.

**Comment by twiecki** [comment]
Then just the doc-string fixes.

**Comment by twiecki** [comment]
Nice!

**Comment by rpgoldman** [comment]
This looks neat!  
I believe that `inference.rst` needs an update, though -- `SMC` isn't in the set of step methods that are included.
The docstring for  `pm.sample` might need some updating, too.  It's actually pretty hard to follow because names are used (especially `nuts`) without being entirely clear about whether they refer to the step function or the initialization method.
I could do the `inference.rst` update, if that would help, but I am not qualified to edit the docstring for `sample`.  I could help review it -- putting my ignorance to good use! -- if someone else would, though.

**Comment by rpgoldman** [comment]
See PR #3514
